src/libraries/Microsoft.PowerFx.Core/Public/Values/FormulaValue.cs

```diff
@@ -58,5 +75,20 @@ public string ToExpression()
 
             return sb.ToString();
         }
+    }
+
+    /// <summary>
+    /// Indicates that a FormulaValue should be copied before being mutated, for Copy on Write semantics.
+    /// </summary>
+    internal interface IMutationCopy
+    {
+        /// <summary>
+        /// Returns a shallow copy of a FormulaValue. For potentially deep data structures such as a Table or Record,
+        /// this includes the head object and any first level collections for rows or fields respectively.
+        /// It stops there, for example even the records within the rows of a Table are not copied.
+        /// If a shallow copy is conditional or not desired, have ShallowCopy "return this;".
```

Comment:
>/// If a shallow copy is conditional or not desired, have ShallowCopy "return this;". [](http://example.com/codeflow?start=8&length=85)

? Could this lead to accidental sharing?
Are callers specifically checked for this?
Would bool TryShallowCopy(out FormulaValue) be more appropriate?

Reply:
We could make this change.  But, it makes the logic for the caller more complicated - they have to check if the copy was made and replace the original only in those cases.  Callers shouldn't need to care if a copy was made.  Perhaps rename it to MaybeShallowCopy?